core/src/main/java/org/testcontainers/containers/DockerComposeContainerInstance.java

```diff
@@ -0,0 +1,141 @@
+package org.testcontainers.containers;
+
+import com.github.dockerjava.api.command.InspectContainerResponse;
+import com.github.dockerjava.api.model.Container;
+import com.github.dockerjava.api.model.ContainerNetwork;
+import com.github.dockerjava.api.model.ExposedPort;
+import com.github.dockerjava.api.model.Ports;
+import com.github.dockerjava.api.model.VolumesFrom;
+import lombok.EqualsAndHashCode;
+import org.testcontainers.containers.wait.WaitStrategy;
+import org.testcontainers.images.RemoteDockerImage;
+import org.testcontainers.utility.DockerImageName;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * Class to provide access to containers started through docker-compose
+ */
+@SuppressWarnings("ConstantConditions")
+@EqualsAndHashCode(callSuper = true)
+public class DockerComposeContainerInstance<SELF extends GenericContainer<SELF>> extends GenericContainer<SELF> {
```

Comment:
Sorry, I'm not sure what this class is doing, could you please explain it a bit more?

Reply:
Hi, sorry, it's probably neither well named nor well documented.

This class represents a service instance that has been created through docker-compose via  `DockerComposeContainer`.

Take for example:

```java
new Docker-compose container(new File("docker-compose.yml")
.withExposedService("redis_1", REDIS_PORT, Wait.forListeningPort())
.withExposedService("elasticsearch_1", ELASTICSEARCH_PORT, Wait.forListeningPort());
```

This will create two `DockerComposeContainerInstances'. One for "redis_" and one for _elasticsearch_1".

It represents the service with the same API as GenericContainer, so things like `getContainerId()`, `getMappedPort()`, etc behave the same way as a `GenericContainer`.

This is why the constructor sets up things by calling setters and with... methods, so that the corresponding getters work the same way as if the service had been created through a `GenericContainer`.

The `DockerComposeContainerInstance` is then passes to the `WaitStrategy` via `waitUntilReady(GenericContainer container)`, and means that any `WaitStrategy` should work the same way for a service created through `DockerComposeContainer` as one created through `GenericContainer`.

Reply:
I see now. So, the problem is that WaitStrategy is highly coupled with GenericContainer, right?

We should probably solve that instead of proxying with "fake" GenericContainer because otherwise, we will have to keep it in sync.
Let me check what can we do with it without breaking the compatibility. 

Maybe, as a first step, we could add GenericContainer-agnostic wait strategies *and* adapt GenericContainer before calling them. WDYT?

Reply:
Yes, `GenericContainer` and `WaitStrategy` are highly coupled and I was conscious of not breaking backwards compatibility.

I thought it might be possible to break out `Container` into separate interfaces, one with just getters and exec methods that represents a container that has already been started, and one with setters for setting up a container before starting it.

The getter interface could then be passed to the `WaitStrategy`.

The problem was that I couldn't find a way to do it without breaking compatibility.

Reply:
Yes, we already have this (separation of container's config and state) in our minds and plan to implement really soon :)

Reply:
It would be pretty straightforward to update existing WaitStrategy classes to use that.

The difficulty comes with users who have implemented their own custom `WaitStrategy` classes.

I'm not sure how common that is, but I had it in mind when implementing this.

Reply:
Yes, me either, but I was thinking to implement it the way their custom strategies will "just work" with GenericContainer but the compiler will not allow them to use with Docker Compose (new feature), so that they will need to update their implementation